### PR TITLE
1953 catch clause

### PR DESCRIPTION
**Pull request by Avuxo**
implements: #1953

### CODE REVIEW

**Comment by ariya** [comment]
> Open to suggestions: because this feature is not implemented in Node.js 6.x and 8.x, these tests cannot properly be run on continuous integration. The tests pass with Node 10.x (because that's when the feature was implemented).

Can you elaborate? The tests should be irrelevant of the run-time Node.js being used.

**Comment by Avuxo** [comment]
The test262 suite _will_ report that non-whitelisted passes were detected if the tests are being run with Node <10. try/catch blocks have not been backported to any prior Node versions yet, so it only works with versions supporting V8 6.6+. Non-bound catch blocks are still a syntax error with older versions of node.

**Comment by ariya** [comment]
> The test262 suite _will_ report that non-whitelisted passes were detected if the tests are being run with Node <10. try/catch blocks have not been backported to any prior Node versions yet, so it only works with versions supporting V8 6.6+. Non-bound catch blocks are still a syntax error with older versions of node.

Oh I think you also need to update the whitelists and check that in as well.

**Comment by JosephPecoraro** [comment]
Is this waiting on something? Being able to parse optional catch binding would be great.

**Comment by Avuxo** [comment]
Unfortunately because there are discrepancies between what will pass/fail on different versions of Node, I can't make CI pass (circle) because it uses a single whitelist file for multiple versions. I'm not really certain this can come out of PR limbo, so if you want the functionality you might just want to use my fork for now.

**Comment by ariya** [comment]
The issue #1953 itself also needs some details. See my comment over there.

**Comment by JosephPecoraro** [comment]
> The issue #1953 itself also needs some details. See my comment over there.

I've added more detail there.

**Comment by JosephPecoraro** [comment]
Seems this needs:

* [ ] Trivially update scanner.ts because tslint complains (`tsfmt --replace --verbose src/*.ts`)
* [ ] Rename paths from `es2018` to `es2019` given ESTree treats it as a 2019 addition
* [ ] Include `test/fixtures/es2019/optional-catch-binding/optional-catch-binding.tree.json`
* [ ] Run `node test/test-262.js --update-whitelist` since `test/language/statements/try/optional-catch-*` now pass.
* [ ] Ensure `npm test` now passes everything

The diff I needed:

```diff
diff --git a/src/scanner.ts b/src/scanner.ts
index d2064f0..bd412a2 100644
--- a/src/scanner.ts
+++ b/src/scanner.ts
@@ -1117,8 +1117,8 @@ export class Scanner {
                 // avoid throwing on regular expressions that are only valid in
                 // combination with the "u" flag.
                 .replace(
-                /[\uD800-\uDBFF][\uDC00-\uDFFF]/g,
-                astralSubstitute
+                    /[\uD800-\uDBFF][\uDC00-\uDFFF]/g,
+                    astralSubstitute
                 );
         }
 
diff --git a/test/test-262-whitelist.txt b/test/test-262-whitelist.txt
index c88c719..4ea65b8 100644
--- a/test/test-262-whitelist.txt
+++ b/test/test-262-whitelist.txt
@@ -384,26 +384,8 @@ test/built-ins/JSON/stringify/bigint.js(default)
 test/built-ins/JSON/stringify/bigint.js(strict mode)
 test/built-ins/Object/setPrototypeOf/bigint.js(default)
 test/built-ins/Object/setPrototypeOf/bigint.js(strict mode)
-test/built-ins/RegExp/named-groups/groups-object.js(default)
-test/built-ins/RegExp/named-groups/groups-object.js(strict mode)
-test/built-ins/RegExp/named-groups/groups-properties.js(default)
-test/built-ins/RegExp/named-groups/groups-properties.js(strict mode)
-test/built-ins/RegExp/named-groups/lookbehind.js(default)
-test/built-ins/RegExp/named-groups/lookbehind.js(strict mode)
-test/built-ins/RegExp/named-groups/non-unicode-malformed.js(default)
-test/built-ins/RegExp/named-groups/non-unicode-malformed.js(strict mode)
-test/built-ins/RegExp/named-groups/non-unicode-match.js(default)
-test/built-ins/RegExp/named-groups/non-unicode-match.js(strict mode)
-test/built-ins/RegExp/named-groups/non-unicode-property-names.js(default)
-test/built-ins/RegExp/named-groups/non-unicode-property-names.js(strict mode)
-test/built-ins/RegExp/named-groups/non-unicode-references.js(default)
-test/built-ins/RegExp/named-groups/non-unicode-references.js(strict mode)
-test/built-ins/RegExp/named-groups/unicode-match.js(default)
-test/built-ins/RegExp/named-groups/unicode-match.js(strict mode)
 test/built-ins/RegExp/named-groups/unicode-property-names.js(default)
 test/built-ins/RegExp/named-groups/unicode-property-names.js(strict mode)
-test/built-ins/RegExp/named-groups/unicode-references.js(default)
-test/built-ins/RegExp/named-groups/unicode-references.js(strict mode)
 test/language/expressions/addition/bigint-arithmetic.js(default)
 test/language/expressions/addition/bigint-arithmetic.js(strict mode)
 test/language/expressions/addition/coerce-bigint-to-string.js(default)
@@ -7385,8 +7367,6 @@ test/language/statements/for-await-of/async-gen-dstr-var-obj-ptrn-rest-val-obj.j
 test/language/statements/for-await-of/async-gen-dstr-var-obj-ptrn-rest-val-obj.js(strict mode)
 test/language/statements/for-await-of/let-block-with-newline.js(default)
 test/language/statements/for-await-of/let-identifier-with-newline.js(default)
-test/annexB/built-ins/RegExp/named-groups/non-unicode-malformed.js(default)
-test/annexB/built-ins/RegExp/named-groups/non-unicode-malformed.js(strict mode)
 test/annexB/language/statements/for-await-of/iterator-close-return-emulates-undefined-throws-when-called.js(default)
 test/annexB/language/statements/for-await-of/iterator-close-return-emulates-undefined-throws-when-called.js(strict mode)
 test/built-ins/BigInt/prototype/toString/radix-2-to-36.js(default)
@@ -7826,14 +7806,6 @@ test/language/statements/try/early-catch-lex.js(default)
 test/language/statements/try/early-catch-lex.js(strict mode)
 test/language/statements/try/early-catch-var.js(default)
 test/language/statements/try/early-catch-var.js(strict mode)
-test/language/statements/try/optional-catch-binding-finally.js(default)
-test/language/statements/try/optional-catch-binding-finally.js(strict mode)
-test/language/statements/try/optional-catch-binding-lexical.js(default)
-test/language/statements/try/optional-catch-binding-lexical.js(strict mode)
-test/language/statements/try/optional-catch-binding-throws.js(default)
-test/language/statements/try/optional-catch-binding-throws.js(strict mode)
-test/language/statements/try/optional-catch-binding.js(default)
-test/language/statements/try/optional-catch-binding.js(strict mode)
 test/language/statements/while/decl-cls.js(default)
 test/language/statements/while/decl-cls.js(strict mode)
 test/language/statements/while/decl-fun.js(default)
```

Then everything looks good! I'll have to go through quite a process to see if I can contribute. If someone wants to pick this up, please be my guest!

File: src/parser.ts [hunk 1]
```diff
@@ -2611,9 +2611,15 @@ export class Parser {
 
     parseCatchClause(): Node.CatchClause {
         const node = this.createNode();
+        let body;
 
         this.expectKeyword('catch');
 
+        if (!this.match('(')) {
+            body = this.parseBlock();
```
Comment (michaelficarra) [inline src/parser.ts]: I would rather not duplicate this part. Just set `param` to `null` and skip the parameter parsing.

File: test/test-262-whitelist.txt [hunk 2]
```diff
@@ -384,40 +384,6 @@ test/built-ins/JSON/stringify/bigint.js(default)
 test/built-ins/JSON/stringify/bigint.js(strict mode)
 test/built-ins/Object/setPrototypeOf/bigint.js(default)
 test/built-ins/Object/setPrototypeOf/bigint.js(strict mode)
-test/built-ins/RegExp/lookBehind/alternations.js(default)
```
Comment (michaelficarra) [inline test/test-262-whitelist.txt]: These changes appear unrelated, and should probably be made in a separate PR.

**Comment by Meir017** [comment]
@Avuxo pr has conflicts :)

**Comment by Meir017** [comment]
#2050 relates to the same feature
